internal/querynode/task_scheduler.go

```diff
@@ -84,9 +121,193 @@ func (s *taskScheduler) taskLoop() {
 func (s *taskScheduler) Start() {
 	s.wg.Add(1)
 	go s.taskLoop()
+
+	s.wg.Add(1)
+	go s.scheduleSQTasks()
+
+	s.wg.Add(1)
+	go s.executeSQTasks()
+}
+
+func (s *taskScheduler) scheduleSQTasks() {
+	defer s.wg.Done()
+	for {
+		select {
+		case <-s.ctx.Done():
+			log.Info("QueryNode sop schedulerSQTasks")
+			return
+		case <-s.notifyChan:
+			log.Info("QueryNode taskScheduler notifyChan")
+			s.tryMergeSQTasks(nil)
+			s.popAndAddToExecute()
+
+		case t, ok := <-s.receiveSQTaskChan:
+			var tasks []sqTask
+			tasks = append(tasks, t)
+			if ok {
+				pendingTaskLen := len(s.receiveSQTaskChan)
+				for i := 0; i < pendingTaskLen; i++ {
+					t := <-s.receiveSQTaskChan
+					tasks = append(tasks, t)
+				}
+				log.Info("QueryNode taskScheduler receiveSQTaskChan")
+				s.tryMergeSQTasks(tasks)
+				s.popAndAddToExecute()
+			} else {
+				errMsg := "taskScheduler receiveSQTaskChan has been closed"
+				log.Warn(errMsg)
+				return
+			}
+		case <-s.tsafeUpdateChan:
+			s.tryMergeSQTasks(nil)
+			s.popAndAddToExecute()
+		}
+	}
+}
+
+func (s *taskScheduler) addSQTask(ctx context.Context, t sqTask) error {
+	select {
+	case <-ctx.Done():
+		return fmt.Errorf("taskScheduler addSQTask context is done")
+	case <-s.ctx.Done():
+		return fmt.Errorf("taskScheduler stoped")
+	case s.receiveSQTaskChan <- t:
+		return nil
+	}
+}
+
+func (s *taskScheduler) popAndAddToExecute() {
+	readyLen := s.readySQTasks.Len()
+	if readyLen <= 0 {
+		return
+	}
+	curUsage := atomic.LoadInt32(&s.cpuUsage)
+	targetUsage := s.maxCPUUsage - curUsage
+	if targetUsage <= 0 {
+		return
+	}
+	tasks, deltaUsage := s.scheduleSQPolicy(s.readySQTasks, targetUsage)
+	if deltaUsage <= 0 || tasks == nil {
+		return
+	}
+	atomic.AddInt32(&s.cpuUsage, deltaUsage)
+	for _, t := range tasks {
+		s.executeSQTaskChan <- t
+	}
+}
+
+func (s *taskScheduler) executeSQTasks() {
+	defer s.wg.Done()
+	var taskWg sync.WaitGroup
+	defer taskWg.Wait()
+	for {
+		select {
+		case <-s.ctx.Done():
+			log.Debug("QueryNode stop executeSQTasks", zap.Int64("NodeID", Params.QueryNodeCfg.GetNodeID()))
+			return
+		case t, ok := <-s.executeSQTaskChan:
+			if ok {
+				taskWg.Add(1)
+				go func(t sqTask) {
+					defer taskWg.Done()
+					s.processSQTask(t)
+					atomic.AddInt32(&s.cpuUsage, -t.EstimateCPUUsage())
```

Comment:
avoid calculation overflow

Reply:
a shared variable  `currentA`

gorountine A:
        atomic.Load(&currentA)
        x = getAPositiveNumber()
        atomic.Add(&currentA,  x)

                                 x
   --->  channel -----> 协程B 

gorountine B:
    atomic.Add(&currentA, -x)

Since there is a channel in the middle, it is guaranteed that the subtraction will always occur after the addition. So it is guaranteed not to generate negative numbers.